### PR TITLE
Resolve #1176

### LINKED ISSUE #1176: Add plan info to planned queries
In order to support adding auxiliary information to the plan, allow plans to reference additional information.

### PR DESCRIPTION
**Pull request by ohadzeliger**
Create QueryPlanResult that holds both QueryRecordPlan as well as QueryPlanInfo.
PlanInfo can be used to hold additional information about the plans (in key-value pairs).

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java [hunk 1]
```diff
@@ -229,7 +230,7 @@ public RecordQueryPlan plan(@Nonnull RecordQuery query) {
                         "explain", PlannerGraphProperty.explain(singleRoot)));
             }
 
-            return (RecordQueryPlan)singleRoot;
+            return new QueryPlanResult((RecordQueryPlan)singleRoot);
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:233]: can we add the task queue size metrics here so this PR serves an immediate purpose?

**Review by normen662 — CHANGES_REQUESTED** [review]
There is a lot of `.getPlan()` in here. Not sure how to not do that, though.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java (hunk 1 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java [hunk 2]
```diff
@@ -67,6 +67,18 @@
     @Nonnull
     RecordQueryPlan plan(@Nonnull RecordQuery query);
 
+    /**
+     * Create a plan to get the results of the provided query.
+     * This method returns a {@link QueryPlanResult} that contains the same plan ass returned by {@link #plan(RecordQuery)}
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java:72]: PG13 :-)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java [hunk 3]
```diff
@@ -237,6 +244,15 @@ public RecordQueryPlan plan(@Nonnull RecordQuery query) {
         }
     }
 
+    @Nonnull
+    @Override
+    public QueryPlanResult planForQuery(@Nonnull final RecordQuery query) {
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:249]: This method's name is not really reflecting what's going on. `planWithInfo` maybe?

**Review by normen662 — CHANGES_REQUESTED** [review]
Can you please add a one-liner also to the output of the debugger's `tasks` command? It would be very useful to have that printed.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java [hunk 4]
```diff
@@ -67,6 +67,18 @@
...
+     * with additional information provided in the {@link QueryPlanInfo}
+     *
+     * @param query a query for records on this planner's metadata
+     * @return a {@link QueryPlanResult} that contains the plan for the query with additional information
+     * @throws com.apple.foundationdb.record.RecordCoreException if the planner cannot plan the query
+     */
+    @Nonnull
+    QueryPlanResult planForQuery(@Nonnull RecordQuery query);
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java:80]: I think the existing `plan` method is a verb. This looks like a noun (if it were a verb, I guess it'd mean prepare something in anticipation of this query). So I'd go with `planQuery` or `getPlanForQuery` or `buildPlanForQuery` of something like that.

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java (hunk 2 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java]: Done

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java (hunk 3 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java (hunk 4 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanner.java]: Done.

**Comment by ohadzeliger** [comment]
test this please

**Review by normen662 — APPROVED** [review]

**Review by MMcM — APPROVED** [review]

**Pull request merged**